crates/air/src/commands/format.rs

```diff
@@ -1,64 +1,312 @@
+        FormatMode::Write => {
+            if errors.is_empty() {
+                Ok(ExitStatus::Success)
+            } else {
+                Ok(ExitStatus::Error)
+            }
+        }
+        FormatMode::Check => {
+            if errors.is_empty() {
+                let any_changed = results.iter().any(FormatFileResult::is_changed);
+
+                if any_changed {
+                    Ok(ExitStatus::Failure)
+                } else {
+                    Ok(ExitStatus::Success)
+                }
+            } else {
+                Ok(ExitStatus::Error)
+            }
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum FormatMode {
+    Write,
+    Check,
 }
 
-// TODO: Should you exit after the first failure? Probably not, probably power through
-// and elegantly report failures at the end? Since you've formatted stuff up to
-// that point already anyways.
-// TODO: Hook this up to a command
-// TODO: Ignore anything but R files, of course
-fn _format_dir(path: &PathBuf) -> anyhow::Result<ExitStatus> {
-    let iter = std::fs::read_dir(path)?;
-
-    for file in iter {
-        let Ok(file) = file else {
-            continue;
-        };
-        format_file(&file.path())?;
+impl FormatMode {
+    fn from_command(command: &FormatCommand) -> Self {
+        if command.check {
+            FormatMode::Check
+        } else {
+            FormatMode::Write
+        }
     }
+}
 
-    Ok(ExitStatus::Success)
+fn write_changed(results: &[FormatFileResult], f: &mut impl Write) -> io::Result<()> {
+    for path in results
+        .iter()
+        .filter_map(|result| match result {
+            FormatFileResult::Formatted(path) => Some(path),
+            FormatFileResult::Unchanged => None,
+        })
+        .sorted_unstable()
+    {
+        writeln!(f, "Would reformat: {}", path.display())?;
+    }
+
+    Ok(())
 }
 
-fn format_file(path: &PathBuf) -> anyhow::Result<ExitStatus> {
-    let contents = std::fs::read_to_string(path)?;
+fn resolve_paths(paths: &[PathBuf]) -> Vec<Result<PathBuf, ignore::Error>> {
+    let paths: Vec<PathBuf> = paths.iter().map(air_fs::normalize_path).collect();
 
-    let line_ending = line_ending::infer(&contents);
+    let (first_path, paths) = paths
+        .split_first()
+        .expect("Clap should ensure at least 1 path is supplied.");
 
-    // Normalize to Unix line endings
-    let contents = match line_ending {
-        LineEnding::Lf => contents,
-        LineEnding::Crlf => line_ending::normalize(contents),
-    };
+    // TODO: Parallel directory visitor
+    let mut builder = ignore::WalkBuilder::new(first_path);
 
-    let parser_options = RParserOptions::default();
-    let parsed = air_r_parser::parse(contents.as_str(), parser_options);
+    for path in paths {
+        builder.add(path);
+    }
 
-    if parsed.has_errors() {
-        return Ok(ExitStatus::Error);
+    let mut out = Vec::new();
+
+    for path in builder.build() {
+        match path {
```

Comment:
We utilize BurntSushi's ignore crate (builds on walkdir) to recursively walk the supplied paths (files or directories) and collect all nested R files 

We will eventually add `exclude` and `include` criteria on top of this

We can also eventually make this a parallel walker for performance, but that is harder and can be its own thing